pkg/detectors/postgres/postgres.go

```diff
@@ -69,186 +139,87 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) ([]dete
 	return results, nil
 }
 
-func getDeadlineInSeconds(ctx context.Context) int {
-	deadline, ok := ctx.Deadline()
-	if !ok {
-		// Context does not have a deadline
-		return 0
-	}
-
-	duration := time.Until(deadline)
-	return int(duration.Seconds())
-}
-
-func findUriMatches(dataStr string) url.URL {
-	var pgURL url.URL
-	for _, uri := range uriPattern.FindAllString(dataStr, -1) {
-		pgURL, err := url.Parse(uri)
+func findUriMatches(data []byte) []map[string]string {
+	var matches []map[string]string
+	for _, uri := range uriPattern.FindAll(data, -1) {
+		connStr, err := pq.ParseURL(string(uri))
 		if err != nil {
 			continue
 		}
-		if pgURL.User != nil {
-			return *pgURL
-		}
-	}
-	return pgURL
-}
-
-// check if postgres is running
-func postgresRunning(hostname, port string) bool {
-	connStr := fmt.Sprintf("host=%s port=%s sslmode=disable", hostname, port)
-	db, err := sql.Open("postgres", connStr)
-	if err != nil {
-		return false
-	}
-	defer db.Close()
-	return true
-}
 
-func findComponentMatches(verify bool, dataStr string) []url.URL {
-	usernameMatches := usernamePattern.FindAllStringSubmatch(dataStr, -1)
-	passwordMatches := passwordPattern.FindAllStringSubmatch(dataStr, -1)
-	hostnameMatches := hostnamePattern.FindAllStringSubmatch(dataStr, -1)
-	portMatches := portPattern.FindAllStringSubmatch(dataStr, -1)
-
-	var pgURLs []url.URL
-
-	hosts := findHosts(verify, hostnameMatches, portMatches)
-
-	for _, username := range dedupMatches(usernameMatches) {
-		for _, password := range dedupMatches(passwordMatches) {
-			for _, host := range hosts {
-				hostname, port := strings.Split(host, ":")[0], strings.Split(host, ":")[1]
-				if combinedLength := len(username) + len(password) + len(hostname); combinedLength > 255 {
-					continue
-				}
-				postgresURL := url.URL{
-					Scheme: "postgresql",
-					User:   url.UserPassword(username, password),
-					Host:   fmt.Sprintf("%s:%s", hostname, port),
-				}
-				pgURLs = append(pgURLs, postgresURL)
-			}
+		params := make(map[string]string)
+		parts := connStrPartPattern.FindAllStringSubmatch(connStr, -1)
```

Comment:
nit: If we construct `params` after parts, we can pre-alloc the map. it'll make it go SOOOO much faster :)

`params := make(map[string]string, len(parts))`